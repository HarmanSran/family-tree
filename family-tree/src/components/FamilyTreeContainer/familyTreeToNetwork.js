const parseId = (id) => parseInt(id, 10);

/**
 * Returns object that maps id to 'level'
 *
 * Level is an integer that determine position (i.e. row) in family tree
 *
 * Level = 1 means bottom of tree (the newest children)
 *
 * Highest level is the oldest ancestor
 *
 * @param {object} json From `tree.json`
 * @returns {object} levels
 */
const getNodeLevels = (json) => {
  const levels = {};

  Object.keys(json).forEach((id) => {
    // If this person is not a parent, they will stay at bottom of tree
    if (levels[id] === undefined) {
      levels[id] = 1;
    }

    // Ensure parents are always one level above children
    levels[json[id].relations.father] = levels[id] + 1;
    levels[json[id].relations.mother] = levels[id] + 1;
  });

  return levels;
};

/**
 * Returns array containing nodes of the family tree
 *
 * Node shape:
 * `id` is a unique key (integer)
 * `level` is required (integer)
 * `details` contains everything about this person from `tree.json` (object)
 *
 * @param {object} json From `tree.json`
 * @returns {array} nodes
 */
const getNodes = (json) => {
  const levels = getNodeLevels(json);
  return Object.keys(json).map((id) => ({ id: parseId(id), level: levels[id], details: json[id] }));
};

/**
 * Returns array containing edges of the family tree
 *
 * Edge shape:
 * `from` is required (integer)
 * `to` is required (integer)
 * `label` is required (string)
 *
 * @param {object} json From `tree.json`
 * @returns {array} edges
 */
const getEdges = (json) => {
  const edges = [];
  // For each person
  Object.keys(json).forEach((id) => {
    // For each of that person's relations
    Object.entries(json[id].relations).map(([relation, relationId]) => {
      // Connect that person to that relation
      edges.push({ from: relationId, to: parseId(id), label: relation });
      return null;
    });
  });
  return edges;
};

/**
 * Maps tree.json shape to Visjs network shape
 */
const familyTreeToNetwork = (json) => ({ nodes: getNodes(json), edges: getEdges(json) });

export default familyTreeToNetwork;
