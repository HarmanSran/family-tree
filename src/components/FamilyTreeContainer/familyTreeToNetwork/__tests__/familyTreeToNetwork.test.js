import familyTreeToNetwork from '../familyTreeToNetwork';
import familyTree from './tree.json';

describe('familyTreeToNetwork', () => {
  it('maps from json data to visjs network shape', () => {
    const network = familyTreeToNetwork(familyTree);
    expect(network).toMatchSnapshot();
  });
});
