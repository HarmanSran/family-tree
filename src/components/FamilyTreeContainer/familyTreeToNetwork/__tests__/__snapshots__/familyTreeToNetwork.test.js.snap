// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`familyTreeToNetwork maps from json data to visjs network shape 1`] = `
Object {
  "edges": Array [
    Object {
      "from": 1,
      "label": "father",
      "to": 0,
    },
    Object {
      "from": 2,
      "label": "mother",
      "to": 0,
    },
    Object {
      "from": 3,
      "label": "father",
      "to": 1,
    },
    Object {
      "from": 4,
      "label": "mother",
      "to": 1,
    },
    Object {
      "from": 2,
      "label": "wife",
      "to": 1,
    },
    Object {
      "from": 5,
      "label": "father",
      "to": 2,
    },
    Object {
      "from": 6,
      "label": "mother",
      "to": 2,
    },
    Object {
      "from": 1,
      "label": "husband",
      "to": 2,
    },
    Object {
      "from": 4,
      "label": "wife",
      "to": 3,
    },
    Object {
      "from": 3,
      "label": "husband",
      "to": 4,
    },
    Object {
      "from": 6,
      "label": "wife",
      "to": 5,
    },
    Object {
      "from": 5,
      "label": "husband",
      "to": 6,
    },
    Object {
      "from": 1,
      "label": "father",
      "to": 7,
    },
    Object {
      "from": 2,
      "label": "mother",
      "to": 7,
    },
  ],
  "nodes": Array [
    Object {
      "details": Object {
        "relations": Object {
          "father": 1,
          "mother": 2,
        },
      },
      "id": 0,
      "level": 1,
    },
    Object {
      "details": Object {
        "relations": Object {
          "father": 3,
          "mother": 4,
          "wife": 2,
        },
      },
      "id": 1,
      "level": 2,
    },
    Object {
      "details": Object {
        "relations": Object {
          "father": 5,
          "husband": 1,
          "mother": 6,
        },
      },
      "id": 2,
      "level": 2,
    },
    Object {
      "details": Object {
        "relations": Object {
          "wife": 4,
        },
      },
      "id": 3,
      "level": 3,
    },
    Object {
      "details": Object {
        "relations": Object {
          "husband": 3,
        },
      },
      "id": 4,
      "level": 3,
    },
    Object {
      "details": Object {
        "relations": Object {
          "wife": 6,
        },
      },
      "id": 5,
      "level": 3,
    },
    Object {
      "details": Object {
        "relations": Object {
          "husband": 5,
        },
      },
      "id": 6,
      "level": 3,
    },
    Object {
      "details": Object {
        "relations": Object {
          "father": 1,
          "mother": 2,
        },
      },
      "id": 7,
      "level": 1,
    },
  ],
}
`;
